examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/OfferingFragment.kt

```diff
@@ -59,76 +85,104 @@ class OfferingFragment : Fragment(), PackageCardAdapter.PackageCardAdapterListen
                     }
-                },
-                { storeTransaction, _ ->
-                    handleSuccessfulPurchase(storeTransaction.orderId)
                 }
-            )
+            }
         } else {
-            Purchases.sharedInstance.purchaseSubscriptionOptionWith(
-                requireActivity(),
-                currentPackage.product,
-                purchaseOption,
-                { error, userCancelled ->
-                    if (!userCancelled) {
-                        showUserError(requireActivity(), error)
-                    }
-                },
-                { storeTransaction, _ ->
-                    handleSuccessfulPurchase(storeTransaction.orderId)
-                })
+            startPurchasePackage(purchaseOption, currentPackage, null)
         }
     }
 
     override fun onPurchaseProductClicked(
         cardView: View,
         currentProduct: StoreProduct,
-        purchaseOption: PurchaseOption?
+        purchaseOption: PurchaseOption?,
+        isUpgrade: Boolean
     ) {
-        binding.purchaseProgress.visibility = View.VISIBLE
-
-        if (purchaseOption == null) {
-            Purchases.sharedInstance.purchaseProductWith(
-                requireActivity(),
-                currentProduct,
-                { error, userCancelled ->
-                    if (!userCancelled) {
-                        showUserError(requireActivity(), error)
+        toggleLoadingIndicator(true)
+
+        if (isUpgrade) {
+            showOldSubIdPicker { subId ->
+                subId?.let {
+                    showProrationModePicker { prorationMode ->
+                        prorationMode?.let {
+                            val upgradeInfo = UpgradeInfo(
+                                subId,
+                                prorationMode
+                            )
+                            startPurchaseProduct(purchaseOption, currentProduct, upgradeInfo)
+                        }
                     }
-                },
-                { storeTransaction, _ ->
-                    handleSuccessfulPurchase(storeTransaction.orderId)
                 }
-            )
+            }
         } else {
-            Purchases.sharedInstance.purchaseSubscriptionOptionWith(
-                requireActivity(),
-                currentProduct,
-                purchaseOption,
-                { error, userCancelled ->
-                    if (!userCancelled) {
-                        showUserError(requireActivity(), error)
-                    }
-                },
-                { storeTransaction, _ ->
-                    handleSuccessfulPurchase(storeTransaction.orderId)
-                })
+            startPurchaseProduct(purchaseOption, currentProduct, null)
+        }
+    }
+
+    private fun startPurchaseProduct(
+        purchaseOption: PurchaseOption?,
+        currentProduct: StoreProduct,
+        upgradeInfo: UpgradeInfo?
+    ) {
+        if (upgradeInfo == null) {
```

Comment:
There's a lot of if else here, this could alternatively be written like:

```
when {
    upgradeInfo != null && purchaseOption != null -> Purchases.sharedInstance.purchaseSubscriptionOptionWith(
        requireActivity(),
        currentProduct,
        purchaseOption,
        upgradeInfo,
        purchaseErrorCallback,
        successfulUpgradeCallback
    )
    upgradeInfo != null && purchaseOption == null -> Purchases.sharedInstance.purchaseProductWith(
        requireActivity(),
        currentProduct,
        upgradeInfo,
        purchaseErrorCallback,
        successfulUpgradeCallback
    )
    upgradeInfo == null && purchaseOption != null -> Purchases.sharedInstance.purchaseSubscriptionOptionWith(
        requireActivity(),
        currentProduct,
        purchaseOption,
        purchaseErrorCallback,
        successfulPurchaseCallback)
    else -> Purchases.sharedInstance.purchaseProductWith(
        requireActivity(),
        currentProduct,
        purchaseErrorCallback,
        successfulPurchaseCallback
    )
}
```

which is not necessarily better but just wanted to offer an alternative :)

Reply:
i like that better! more compact. thanks :)